vm/rust/src/juno_state_reader.rs

```diff
@@ -251,7 +272,106 @@ fn native_try_from_json_string(
     //   2. Refactoring the code on the Cairo mono-repo
 
     let sierra_program = sierra_contract_class.extract_sierra_program()?;
-    let executor = compile_and_load(&sierra_program)?;
+
+    // todo(xrvdg) lift this match out of the function once we do not need sierra_program anymore
+    let executor = match load_compiled_contract(&sierra_program, library_output_path) {
+        Some(executor) => {
+            executor.or_else(|_err| compile_and_load(sierra_program, library_output_path))
+        }
+        None => compile_and_load(sierra_program, library_output_path),
+    }?;
 
     Ok(NativeContractClassV1::new(executor, sierra_contract_class)?)
 }
+
+/// Load a contract that is already compiled.
+///
+/// Returns None if the contract does not exist at the output_path.
+///
+/// To compile and load a contract use [persist_from_native_module] instead.
+fn load_compiled_contract(
+    sierra_program: &Program,
+    library_output_path: &PathBuf,
+) -> Option<Result<AotNativeExecutor, Box<dyn std::error::Error>>> {
```

Comment:
Just general feedback in case you didn't know, there is a library for returning dynamic errors: `anyhow` not important for this PR just sharing in case you didn't know